app/src/lib/branch.ts

```diff
@@ -1,25 +1,41 @@
 import { Branch, BranchType } from '../models/branch'
 import { UpstreamRemoteName } from './stores'
+import {
+  RepositoryWithGitHubRepository,
+  getNonForkGitHubRepository,
+} from '../models/repository'
 
 /**
- * Finds the remote branch for a branch in the upstream repository
+ * Finds the default branch of the upstream repository of the passed repository.
  *
- * For example:
- * If official/funnel has the branch `development`, then running
- * this on the branches from the fork outofambit/funnel with
- * `development` as the specified branch name will find the
- * branch `remotes/upstream/development`
+ * When the passed repository is not a fork or the fork is configured to use itself
+ * as the ForkContributionTarget, then this method will return null. Otherwise it'll
+ * return the default branch of the upstream repository.
  *
- * @param branchName short name of the branch in the upstream repo
- * @param branches all the branches in the local repo
+ * @param repository The repository to use.
+ * @param branches all the branches in the local repo.
  */
-export function findUpstreamRemoteBranch(
-  branchName: string,
+export function findDefaultUpstreamBranch(
+  repository: RepositoryWithGitHubRepository,
   branches: ReadonlyArray<Branch>
 ) {
+  const githubRepository = getNonForkGitHubRepository(repository)
+
+  // This is a bit hacky... we checked if the result of calling
+  // getNonForkGitHubRepository() is the same as the associated
+  // gitHubRepository. When this happens, we know that either the
+  // repository is not a fork or that it's a fork with
+  // ForkContributionTarget=Self.
+  //
+  // TODO: Make this method return the default branch of the
+  // origin repository instead of null in this scenario.
+  if (githubRepository === repository.gitHubRepository) {
+    return null
```

Comment:
Is there value in being able to differentiate between this scenario and when the default branch could not be found? I.e. we return `null` here but `undefined` below from the `.find`. In general I find I'm often confused when I encounter methods with tri-state returns like this so if there's value in callers being able to differentiate between this then I think it should be clearly documented in the function documentation, if not we should return `undefined` here or `null` below.

```suggestion
    return undefined
```

Reply:
Thanks for the suggestion! yup there's no value doing so, it was just muscle memory of returning `null` without realizing that lower down the function there's `find()` that can return `undefined`.

I think in this specific case I'd rather leave this `return null` statement and add a check on `find()` to always return `null` (since anyways the only user of this method is casting the `undefined` to `null`). 

I think these kind of issues could get reduced substantially if we enforced typing the return type of functions (this could also potentially highlight some issues with our types where sometimes TypeScript infers `any` somewhere and this gets returned in a function.

Reply:
> I think in this specific case I'd rather leave this return null statement and add a check on find() to always return null (since anyways the only user of this method is casting the undefined to null).

Yep, that's fine with me!

> I think these kind of issues could get reduced substantially if we enforced typing the return type of functions

I think you're right but it's also a tradeoff, most of the time it's really convenient to be able to rely on the inference. I wouldn't want to mandate explicit typing everywhere but there's certainly lots of places where adding it as a form of codified expectation of the method is a great idea and I've never rejected a contribution for using explicitly typed return params.